File: imu_calib_python/imu_calib/utils/detect_static_segments.py
from __future__ import annotations

import numpy as np

from imu_calib.runtime.default_calib_options import default_calib_options
from imu_calib.utils.exceptions import ImuCalibError
from imu_calib.utils.math_utils import as_matrix_n3, validate_time_vector


def detect_static_segments(
    t: np.ndarray,
    gyro: np.ndarray,
    acc: np.ndarray,
    *,
    static_window_sec: float | None = None,
    gyro_norm_threshold: float | None = None,
    gyro_std_threshold: float | None = None,
    acc_std_threshold: float | None = None,
    min_segment_sec: float | None = None,
) -> dict:
    """Detect static segments from gyro and accelerometer data."""
    t = validate_time_vector(t, "t")
    gyro = as_matrix_n3(gyro, "gyro")
    acc = as_matrix_n3(acc, "acc")
    if gyro.shape[0] != t.size or acc.shape[0] != t.size:
        raise ImuCalibError("t, gyro, and acc must have matching lengths.")

    defaults = default_calib_options()["segmentation"]
    dt = float(np.median(np.diff(t)))
    static_window_sec = defaults["static_window_sec"] if static_window_sec is None else float(static_window_sec)
    gyro_norm_threshold = defaults["gyro_norm_threshold"] if gyro_norm_threshold is None else float(gyro_norm_threshold)
    gyro_std_threshold = defaults["gyro_std_threshold"] if gyro_std_threshold is None else float(gyro_std_threshold)
    acc_std_threshold = defaults["acc_std_threshold"] if acc_std_threshold is None else float(acc_std_threshold)
    min_segment_sec = defaults["min_segment_sec"] if min_segment_sec is None else float(min_segment_sec)

    win = max(3, int(round(static_window_sec / dt)))
    gyro_norm = np.linalg.norm(gyro, axis=1)
    gyro_std = _moving_std_scalar(gyro_norm, win)
    acc_mag = np.linalg.norm(acc, axis=1)
    acc_std = _moving_std_scalar(acc_mag, win)
    mask = (gyro_norm <= gyro_norm_threshold) & (gyro_std <= gyro_std_threshold) & (acc_std <= acc_std_threshold)
    segments = _logical_mask_to_segments(mask, max(1, int(round(min_segment_sec / dt))))
    return {
        "mask": mask,
        "segments": segments,
        "quality": {
            "window_samples": win,
            "gyro_norm_threshold": gyro_norm_threshold,
            "gyro_std_threshold": gyro_std_threshold,
            "acc_std_threshold": acc_std_threshold,
            "num_segments": int(segments.shape[0]),
        },
    }


def _moving_std_scalar(x: np.ndarray, win: int) -> np.ndarray:
    out = np.zeros_like(x, dtype=float)
    half = win // 2
    for i in range(x.size):
        i0 = max(0, i - half)
        i1 = min(x.size, i + half + 1)
        out[i] = float(np.std(x[i0:i1], ddof=0))
    return out


def _logical_mask_to_segments(mask: np.ndarray, min_len: int) -> np.ndarray:
    d = np.diff(np.r_[False, mask.astype(bool), False].astype(int))
    starts = np.flatnonzero(d == 1)
    ends = np.flatnonzero(d == -1) - 1
    keep = (ends - starts + 1) >= min_len
    return np.column_stack([starts[keep], ends[keep]]) if np.any(keep) else np.empty((0, 2), dtype=int)

File: imu_calib_matlab/src/analysis/detect_static_segments.m
function result = detect_static_segments(t, gyro, acc, varargin)
%DETECT_STATIC_SEGMENTS Detect static segments from gyro and accelerometer data.
% Usage:
%   result = detect_static_segments(t, gyro, acc)
%
% Output:
%   result.mask         : [N x 1] logical
%   result.segments     : [M x 2] start/end indices
%   result.quality      : struct with thresholds and summary

validateattributes(t, {'numeric'}, {'column', 'finite', 'nonempty'}, mfilename, 't', 1);
validateattributes(gyro, {'numeric'}, {'2d', 'ncols', 3, 'finite', 'nonempty'}, mfilename, 'gyro', 2);
validateattributes(acc, {'numeric'}, {'2d', 'ncols', 3, 'finite', 'nonempty'}, mfilename, 'acc', 3);

if size(gyro, 1) ~= numel(t) || size(acc, 1) ~= numel(t)
    error('detect_static_segments:LengthMismatch', ...
        't, gyro, and acc must have matching lengths.');
end

opts = parse_options(varargin{:});
dt = median(diff(t));
win = max(3, round(opts.static_window_sec / dt));

gyroNorm = sqrt(sum(gyro .^ 2, 2));
gyroStd = moving_std_scalar(gyroNorm, win);
accMag = sqrt(sum(acc .^ 2, 2));
accStd = moving_std_scalar(accMag, win);

mask = (gyroNorm <= opts.gyro_norm_threshold) & ...
       (gyroStd <= opts.gyro_std_threshold) & ...
       (accStd <= opts.acc_std_threshold);

segments = logical_mask_to_segments(mask, round(opts.min_segment_sec / dt));

result = struct();
result.mask = mask;
result.segments = segments;
result.quality = struct();
result.quality.window_samples = win;
result.quality.gyro_norm_threshold = opts.gyro_norm_threshold;
result.quality.gyro_std_threshold = opts.gyro_std_threshold;
result.quality.acc_std_threshold = opts.acc_std_threshold;
result.quality.num_segments = size(segments, 1);
end

function opts = parse_options(varargin)
defaults = default_calib_options();
opts = defaults.segmentation;

if mod(numel(varargin), 2) ~= 0
    error('detect_static_segments:InvalidInput', ...
        'Optional arguments must be provided as name/value pairs.');
end

for k = 1:2:numel(varargin)
    name = lower(char(varargin{k}));
    value = varargin{k + 1};
    switch name
        case 'static_window_sec'
            opts.static_window_sec = double(value);
        case 'gyro_norm_threshold'
            opts.gyro_norm_threshold = double(value);
        case 'gyro_std_threshold'
            opts.gyro_std_threshold = double(value);
        case 'acc_std_threshold'
            opts.acc_std_threshold = double(value);
        case 'min_segment_sec'
            opts.min_segment_sec = double(value);
        otherwise
            error('detect_static_segments:UnknownOption', ...
                'Unknown option "%s".', char(varargin{k}));
    end
end
end

function s = moving_std_scalar(x, win)
s = zeros(size(x));
halfWin = floor(win / 2);
for i = 1:numel(x)
    i0 = max(1, i - halfWin);
    i1 = min(numel(x), i + halfWin);
    s(i) = std(x(i0:i1));
end
end

function segments = logical_mask_to_segments(mask, minLen)
d = diff([false; mask(:); false]);
starts = find(d == 1);
ends = find(d == -1) - 1;
keep = (ends - starts + 1) >= minLen;
segments = [starts(keep), ends(keep)];
end
